### ISSUE TITLE
Conv3d  depthwise is slower than conv2d depthwise  + for loop

### ISSUE THREAD

**Issue by ducha-aiki** [pytorch/pytorch#34089]
## 🐛 Bug

Conv3d depthwise  seems much slower, at least for small number of channels than conv2d depthwise  + for loop

Timings:

> benchmark_conv3d_groups cuda:0
> 10 loops, best of 3: 62.8 ms per loop
> 3 benchmark_conv3d_reshape cuda:0
> 10 loops, best of 3: 68.2 ms per loop
> 3 benchmark_conv2d_cat cuda:0
> 100 loops, best of 3: 9.71 ms per loop
>    
> benchmark_conv3d_groups cpu
> 1 loop, best of 3: 381 ms per loop
> 3 benchmark_conv3d_reshape cpu
> 1 loop, best of 3: 367 ms per loop
> 3 benchmark_conv2d_cat cpu
> 10 loops, best of 3: 172 ms per loop
>    
> benchmark_conv3d_groups cuda:0
> 10 loops, best of 3: 67.1 ms per loop
> 5 benchmark_conv3d_reshape cuda:0
> 10 loops, best of 3: 68.4 ms per loop
> 5 benchmark_conv2d_cat cuda:0
> 100 loops, best of 3: 9.74 ms per loop
>    
> benchmark_conv3d_groups cpu
> 1 loop, best of 3: 916 ms per loop
> 5 benchmark_conv3d_reshape cpu
> 1 loop, best of 3: 339 ms per loop
> 5 benchmark_conv2d_cat cpu
> 10 loops, best of 3: 170 ms per loop

   
## To Reproduce
Run this code (or this colab)
https://colab.research.google.com/drive/1HL3VqE5AH_ZlHT2SQfIub2AVjwdTor5U

```python
import torch
import random
import torch.nn.functional as F
import torch.backends.cudnn as cudnn
import timeit
def benchmark_conv2d_cat(input, device, sh):
    b1,c1,h1,w1 = input.size()
    torch.cuda.synchronize()
    spatial_pad = [kernel.size(1) // 2,
                   kernel.size(1) // 2,
                   kernel.size(2) // 2,
                   kernel.size(2) // 2]
    inp = F.pad(input, spatial_pad, 'replicate')
    out_channels: int = 3 if kernel.size(0) == 3 else 2
    x = torch.cat([F.conv2d(inp, kernel[i:i+1].expand(c1,-1,-1,-1), 
                            groups=c1).unsqueeze(2) for i in range(out_channels)],
                  dim=2)
                   
    loss = x.sum()
    loss.backward()
    torch.cuda.synchronize()
    return


def benchmark_conv3d_reshape(input, device, sh):
    b1,c1,h1,w1 = input.size()
    torch.cuda.synchronize()
    kernel1: torch.Tensor = kernel.unsqueeze(1).unsqueeze(1)

    # convolve input tensor with sobel kernel
    # Pad with "replicate for spatial dims, but with zeros for channel
    spatial_pad = [kernel.size(1) // 2,
                   kernel.size(1) // 2,
                   kernel.size(2) // 2,
                   kernel.size(2) // 2]
    # print (spatial_pad)
    out_channels: int = 3 if kernel.size(0) == 3 else 2
    padded_inp: torch.Tensor = F.pad(input.reshape(b1 * c1, 1, h1, w1),
                                     spatial_pad, 'replicate')[:, :, None]
    #print (padded_inp.shape)
    x =  F.conv3d(padded_inp, kernel1, padding=0).view(b1, c1, out_channels, h1, w1)
    loss = x.sum()
    loss.backward()
    torch.cuda.synchronize()
    return


def benchmark_conv3d_groups(input, device, sh):
    b1,c1,h1,w1 = input.size()
    torch.cuda.synchronize()
    kernel1: torch.Tensor = kernel.unsqueeze(1).unsqueeze(1)

    # Pad with "replicate for spatial dims, but with zeros for channel
    spatial_pad = [kernel.size(1) // 2,
                   kernel.size(1) // 2,
                   kernel.size(2) // 2,
                   kernel.size(2) // 2]
    # print (spatial_pad)
    out_channels: int = 3 if kernel.size(0) == 3 else 2
    padded_inp: torch.Tensor = F.pad(input, spatial_pad, 'replicate')[:, :, None]
    #print (padded_inp.shape)
    #print (kernel_flip.shape)
    x =  F.conv3d(padded_inp, kernel1.repeat(c1,1,1,1,1),
                                                 groups=c1,
                  padding=0).view(b1, c1, out_channels, h1, w1)
    loss = x.sum()
    loss.backward()
    torch.cuda.synchronize()
    return

b,c,h,w = 4,8,512,512
sh = [b,c,h,w]

for ks in [3,5]:
    for device in [torch.device('cuda:0'), torch.device('cpu')]: 
        data = torch.rand(b,c,h,w, device=device).float() 
        b1,c1,h1,w1 = data.size() 
        #warmup 
        kernel = torch.rand(ks//2+1,ks,ks, device=device, requires_grad=True).float() 
        benchmark_conv3d_reshape(data, device,sh)
        benchmark_conv2d_cat(data, device,sh)
        benchmark_conv3d_groups(data, device,sh)
        torch.cuda.synchronize() 
        print ('benchmark_conv3d_groups', str(device)) 
        %timeit benchmark_conv3d_groups(data, device, sh) 
        kernel = torch.rand(2,3,3, device=device, requires_grad=True).float() 
        torch.cuda.synchronize() 
        print (ks, 'benchmark_conv3d_reshape',str(device))
        kernel = torch.rand(2,3,3, device=device, requires_grad=True).float() 
        %timeit benchmark_conv3d_reshape(data, device, sh) 
        kernel = torch.rand(2,3,3, device=device, requires_grad=True).float() 
        print (ks, 'benchmark_conv2d_cat',str(device))
        %timeit benchmark_conv2d_cat(data, device, sh) 
        print ("   ")
```

## Expected behavior

Well, not be  slower than for loop. Although it is probably cudnn issue

## Environment

Google colab GPU

Please copy and paste the output from our

## Additional context
 
Discovered when benchmarked spatial gradient function for kornia
https://github.com/kornia/kornia/issues/466

cc @VitalyFedyunin @ngimel

**Comment by ptrblck** [comment]
Based on your code snippet, it seems you are not using a 3-dimensional convolution, but rather a grouped 2-dimensional convolution.

Note that you create a "fake depth dim" in your `nn.Conv3d`, which is never used.

The fastest approach should therefore be a grouped `nn.Conv2d` with the interleaved kernel (which is created in your for loop):
```python
def benchmark_conv2d_interleaved(input, device, sh):
    torch.cuda.synchronize()
    # interleave kernel
    k_tmp = torch.cat((kernel[0:1].repeat(8, 1, 1, 1), kernel[1:2].repeat(8, 1, 1, 1)), dim=1)
    k_tmp = k_tmp.view(-1, 1, kernel.size(1), kernel.size(2))
    
    spatial_pad = [kernel.size(1) // 2,
                   kernel.size(1) // 2,
                   kernel.size(2) // 2,
                   kernel.size(2) // 2]
    inp = F.pad(input, spatial_pad, 'replicate')
    x =  F.conv2d(inp, k_tmp, groups=8)
    x = x.view(4, 8, 2, 512, 512)
                   
    loss = x.sum()
    loss.backward()
    torch.cuda.synchronize()
    return x
```
Note that you might need to relax the shapes for your use case as I've hardcoded them for profiling `ks==3`.

Also, the `conv3d_reshape` code will be faster when using `nn.Conv2d` and removing the unnecessary depth dimension:
```python
def benchmark_conv2d_reshape(input, device, sh):
    b1,c1,h1,w1 = input.size()
    torch.cuda.synchronize()
    kernel1: torch.Tensor = kernel.unsqueeze(1)

    # convolve input tensor with sobel kernel
    # Pad with "replicate for spatial dims, but with zeros for channel
    spatial_pad = [kernel.size(1) // 2,
                   kernel.size(1) // 2,
                   kernel.size(2) // 2,
                   kernel.size(2) // 2]
    # print (spatial_pad)
    out_channels: int = 3 if kernel.size(0) == 3 else 2
    padded_inp: torch.Tensor = F.pad(input.reshape(b1 * c1, 1, h1, w1),
                                     spatial_pad, 'replicate')
    #print (padded_inp.shape)
    x =  F.conv2d(padded_inp, kernel1, padding=0).view(b1, c1, out_channels, h1, w1)
    loss = x.sum()
    loss.backward()
    torch.cuda.synchronize()
    return x
```

Using these methods, I get:
```python
benchmark_conv2d_interleaved
4.65 ms ± 23.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
benchmark_conv2d_reshape
5.39 ms ± 9.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
benchmark_conv3d_groups cuda:0
10.5 ms ± 19 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
benchmark_conv3d_reshape cuda:0
10.8 ms ± 5.44 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
benchmark_conv2d_cat cuda:0
8.23 ms ± 5.22 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

PS: I also verified the results via:
```python
x0 = benchmark_conv2d_interleaved(data, device, sh)
x1 = benchmark_conv2d_reshape(data, device, sh)
x2 = benchmark_conv3d_reshape(data, device,sh)
x3 = benchmark_conv2d_cat(data, device,sh)
x4 = benchmark_conv3d_groups(data, device,sh)

print(torch.allclose(x0, x1) and torch.allclose(x1, x2) and torch.allclose(x2, x3) and torch.allclose(x3, x4))
```

Let me know, if I misunderstood your use case.

**Comment by ducha-aiki** [comment]
@ptrblck thank you! Indeed, you are right in that sense that my convolution is not really 3d, but I was under the impression that similar to [1xK] kernels in inception-like architectures, which are in fact 1D convolutions, but typically run like 2D convs with extra dim, conv3d should work here as well.
Thank you for the conv2d_interleaved, it solves the problem.

**Comment by ngimel** [comment]
@ptrblck workaround is very clever, but it is still a real bug, in a sense that cudnn should determine under the hood that the convolution is in fact 2d and use available efficient implementation. @ptrblck can you please file a cudnn bug for this?
